zilliqa/src/consensus.rs

```diff
@@ -439,23 +444,30 @@ impl Consensus {
     pub fn download_blocks_up_to_head(&mut self) -> Result<()> {
         let head_block = self.head_block();
 
-        let random_validator_peer = {
-            // Remove self from potential people to send to
-            let mut other_peers = head_block.committee.clone();
-            other_peers.remove_by_peer_id(self.peer_id());
-
-            match other_peers.len() {
-                0 => None,
-                _ => Some(other_peers.choose_random().peer_id),
-            }
-        };
-
+        let random_peer = self.get_random_other_peer();
         self.block_store
-            .request_blocks(random_validator_peer, head_block.header.number + 1)?;
+            .request_blocks(random_peer, head_block.header.number + 1)?;
 
         Ok(())
     }
 
+    pub fn get_random_other_peer(&mut self) -> Option<PeerId> {
+        // Remove self from potential people to send to
```

Comment:
Can you do something like

```rust
use rand::seq::IteratorRandom;
...
self.head_block().committee.iter().filter(|v| v.peer_id != self.peer_id()).choose(&mut self.rng);
```
which will avoid cloning the committee?

The length check should also be unnecessary since `choose` will return `None` if the iterator is empty.

Reply:
Great suggestion, I have implemented it :

```
    pub fn get_random_other_peer(&mut self) -> Option<PeerId> {
        let our_id = self.peer_id();
        self.head_block()
            .committee
            .iter()
            .filter(|v| v.peer_id != our_id)
            .choose(&mut self.rng)
            .map(|v| v.peer_id)
    }
```

`our_id` is captured outside the filter to please the borrow checker.